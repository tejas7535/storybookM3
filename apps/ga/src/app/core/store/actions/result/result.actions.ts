import { createAction, props, union } from '@ngrx/store';

export const getCalculation = createAction('[Result] Get Calculation');

export const calculationSuccess = createAction(
  '[Result] Get Calculation Success',
  props<{ resultId: string }>()
);

export const calculationError = createAction('[Result] Get Calculation Error');

const all = union({
  getCalculation,
  calculationSuccess,
  calculationError,
});

export type CalculationActions = typeof all;
