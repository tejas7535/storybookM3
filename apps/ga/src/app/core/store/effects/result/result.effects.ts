import { Injectable } from '@angular/core';

import { catchError, filter, map, mergeMap, of } from 'rxjs';

import { Actions, concatLatestFrom, createEffect, ofType } from '@ngrx/effects';
import { ROUTER_NAVIGATED } from '@ngrx/router-store';
import { Store } from '@ngrx/store';

import { ApplicationInsightsService } from '@schaeffler/application-insights';

import { GreaseCalculationPath } from '../../../../grease-calculation/grease-calculation-path.enum';
import { PROPERTIES } from '../../../../shared/constants';
import { ErrorService, RestService } from '../../../services';
import {
  calculationError,
  calculationSuccess,
  getCalculation,
} from '../../actions/result/result.actions';
import { getModelId } from '../../selectors/bearing/bearing.selector';
import { getCalculationParameters } from '../../selectors/parameter/parameter.selector';

@Injectable()
export class ResultEffects {
  router$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(ROUTER_NAVIGATED),
      map((action: any) => action.payload.routerState.url),
      map((url: string) =>
        Object.values(GreaseCalculationPath).find(
          (route: string) => route !== '' && url.includes(route)
        )
      ),
      filter(
        (currentRoute: string) =>
          currentRoute === GreaseCalculationPath.ResultPath
      ),
      map(() => getCalculation())
    );
  });

  calculation$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(getCalculation),
      concatLatestFrom(() => this.store.select(getModelId)),
      map(([_action, modelId]) => modelId),
      mergeMap((modelId) =>
        this.restService.getGreaseCalculation(modelId).pipe(
          map((resultId) => calculationSuccess({ resultId })),
          catchError((_e) => of(calculationError()))
        )
      )
    );
  });

  calculationError$ = createEffect(
    () => {
      return this.actions$.pipe(
        ofType(calculationError),
        map(() => this.errorService.openGenericSnackBar())
      );
    },
    { dispatch: false }
  );

  calculationSuccess$ = createEffect(
    () => {
      return this.actions$.pipe(
        ofType(calculationSuccess),
        concatLatestFrom(() => this.store.select(getCalculationParameters)),
        map(([_action, parameters]) => parameters),
        map((properties) =>
          this.applicationInsightsService.logEvent(
            PROPERTIES,
            properties.options
          )
        )
      );
    },
    { dispatch: false }
  );

  constructor(
    private readonly actions$: Actions,
    private readonly restService: RestService,
    private readonly store: Store,
    private readonly errorService: ErrorService,
    private readonly applicationInsightsService: ApplicationInsightsService
  ) {}
}
