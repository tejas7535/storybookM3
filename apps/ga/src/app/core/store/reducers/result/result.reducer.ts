import { Action, createReducer, on } from '@ngrx/store';

import {
  calculationError,
  calculationSuccess,
  getCalculation,
} from '../../actions/result/result.actions';

export interface ResultState {
  resultId: string;
  jsonResult: string;
  htmlResult: string;
  loading: boolean;
}

export const initialState: ResultState = {
  resultId: undefined,
  jsonResult: undefined,
  htmlResult: undefined,
  loading: false,
};

export const resultReducer = createReducer(
  initialState,
  on(
    getCalculation,
    (state: ResultState): ResultState => ({
      ...state,
      loading: true,
    })
  ),
  on(
    calculationSuccess,
    (state: ResultState, { resultId }): ResultState => ({
      ...state,
      resultId,
      loading: false,
    })
  ),
  on(
    calculationError,
    (state: ResultState): ResultState => ({
      ...state,
      loading: false,
    })
  )
);

export function reducer(state: ResultState, action: Action): ResultState {
  return resultReducer(state, action);
}
