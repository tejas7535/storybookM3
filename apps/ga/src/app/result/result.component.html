<ng-container *transloco="let t; read: 'report'">
  <schaeffler-subheader
    class="shrink flex"
    [showBackButton]="true"
    [title]="t('title')"
    (backButtonClicked)="navigateBack()"
  >
  </schaeffler-subheader>
  <div class="py-4">
    <mat-slide-toggle
      color="primary"
      labelPosition="before"
      [(ngModel)]="showCompactView"
    >
      <span class="text-body-1 text-medium-emphasis">
        {{ t('showCompactView') }}
      </span>
    </mat-slide-toggle>
  </div>
  <schaeffler-report
    *ngIf="showCompactView"
    [jsonReport]="jsonReportUrl"
    [type]="reportType"
  ></schaeffler-report>
  <ng-container *ngIf="!showCompactView">
    <schaeffler-report
      *ngIf="reportUrls$ | ngrxPush as reportUrls; else loading_spinner"
      [htmlReport]="reportUrls.htmlReportUrl"
      [reportSelector]="reportSelector"
    >
      <!-- <div class="flex flex-col md:flex-row">
      <a
        mat-raised-button
        color="primary"
        class="grow"
        [href]="result.pdfReportUrl"
      >
        {{ t('downloadReport') }}
      </a>
    </div> -->
    </schaeffler-report>
  </ng-container>
  <ng-template #loading_spinner>
    <schaeffler-loading-spinner></schaeffler-loading-spinner>
  </ng-template>
</ng-container>
