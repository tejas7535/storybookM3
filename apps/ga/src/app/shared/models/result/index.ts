export * from './report-urls.model';
export * from './result.model';
