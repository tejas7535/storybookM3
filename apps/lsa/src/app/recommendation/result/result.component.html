@if (validResult) {
  <div
    class="grid grid-cols-1 min-[1320px]:grid-cols-2 flex-col lg:flex-row gap-4 md:justify-between"
    *transloco="let t; read: 'recommendation.result'"
  >
    <div class="flex flex-col space-y-1 min-[1320px]:col-span-2">
      @for (notice of validResult.notices; track notice.titleId) {
        <div class="rounded-md border-info bg-info text-info px-4 py-2">
          {{ t(notice.titleId) }}
        </div>
      }
      @for (warning of validResult.warnings; track warning.titleId) {
        <div
          class="rounded-md border-warning bg-warning text-warning px-4 py-2"
        >
          {{ t(warning.titleId) }}
        </div>
      }
    </div>
    <div class="flex flex-col gap-4 xl:mr-auto">
      <lsa-recommendation-table
        [data]="validResult.lubricators | lsaRecommendationTableData"
        (recommendedSelectedChange)="onRecommendedSelectedChange($event)"
      ></lsa-recommendation-table>
    </div>
    <div class="order-3 min-[1320px]:order-[0] row-span-3 flex flex-col">
      <h2 class="text-title-small">
        {{
          t('heading', {
            designation: isRecommendedSelected
              ? validResult.lubricators.recommendedLubricator.name
              : validResult.lubricators.minimumRequiredLubricator.name
          })
        }}
      </h2>
      <lsa-accessory-table
        [accessories]="
          isRecommendedSelected
            ? validResult.lubricators.recommendedLubricator.bundle
            : validResult.lubricators.minimumRequiredLubricator.bundle
        "
        [classPriorities]="validResult.classes"
      />
      <div
        class="w-full flex flex-col md:flex-row justify-end space-y-2 md:space-y-0 md:space-x-4 mt-4"
      >
        <button
          (click)="resetForm()"
          mat-icon-button
          class="inline-flex flex-row items-center"
        >
          <mat-icon>restart_alt</mat-icon>
        </button>
        <lsa-add-to-cart-button
          [label]="t('addToCart')"
          [snackbarMessage]="t('addToCartMessage')"
          (addToCart)="onAddToCart()"
        ></lsa-add-to-cart-button>
      </div>
    </div>
    <ng-content></ng-content>
  </div>
} @else {
  <lsa-error-container [response]="errorInstance" />
}
