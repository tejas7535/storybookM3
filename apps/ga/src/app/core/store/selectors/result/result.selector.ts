import { createSelector } from '@ngrx/store';

import { getResultState } from '../../reducers';
import { ResultState } from '../../reducers/result/result.reducer';

export const getResult = createSelector(
  getResultState,
  (state: ResultState): ResultState => state
);

export const getResultId = createSelector(
  getResultState,
  (state: ResultState): string => state?.resultId
);
