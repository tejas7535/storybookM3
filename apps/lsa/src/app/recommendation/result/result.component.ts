import {
  Component,
  Input,
  OnChanges,
  SimpleChanges,
  ViewChild,
} from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatIcon, MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';

import { AddToCartService } from '@lsa/core/services/add-to-cart.service';
import { ErrorResponse, RecommendationResponse } from '@lsa/shared/models';
import { RecommendationTableDataPipe } from '@lsa/shared/pipes/recommendation-table-data.pipe';

import { SharedTranslocoModule } from '@schaeffler/transloco';

import { AccessoryTableComponent } from './accessory-table/accessory-table.component';
import { AddToCartButtonComponent } from './add-to-cart-button/add-to-cart-button.component';
import { RecommendationTableComponent } from './recommendation-table/recommendation-table.component';

@Component({
  selector: 'lsa-result',
  standalone: true,
  imports: [
    RecommendationTableComponent,
    RecommendationTableDataPipe,
    MatIcon,
    MatFormFieldModule,
    MatInputModule,
    AccessoryTableComponent,
    MatButtonModule,
    MatIconModule,
    SharedTranslocoModule,
    AddToCartButtonComponent,
  ],
  templateUrl: './result.component.html',
})
export class ResultComponent implements OnChanges {
  @Input() recommendationResult: RecommendationResponse | ErrorResponse;

  @ViewChild(AccessoryTableComponent)
  accessoryTableComponent: AccessoryTableComponent;

  isRecommendedSelected = true;
  validResult?: RecommendationResponse;
  errorInstance: ErrorResponse;

  constructor(private readonly addToCartService: AddToCartService) {}

  onRecommendedSelectedChange(isRecommendedSelected: boolean): void {
    this.isRecommendedSelected = isRecommendedSelected;
  }

  ngOnChanges(_changes: SimpleChanges): void {
    if ('name' in this.recommendationResult) {
      // TODO: Build error handling logic
      this.errorInstance = this.recommendationResult as ErrorResponse;
      this.validResult = undefined;
    } else {
      this.validResult = this.recommendationResult as RecommendationResponse;
      this.isRecommendedSelected =
        !!this.validResult.lubricators.recommendedLubricator;
    }
  }

  onAddToCart(): void {
    this.addToCartService.addToCartEvent(
      [...this.accessoryTableComponent.accessories],
      this.accessoryTableComponent.tableFormGroup
    );
  }
}
