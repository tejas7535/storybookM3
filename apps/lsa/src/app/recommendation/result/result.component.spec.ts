import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

import { provideTranslocoTestingModule } from '@schaeffler/transloco/testing';

import { ResultComponent } from './result.component';

describe('ResultComponent', () => {
  let component: ResultComponent;
  let spectator: Spectator<ResultComponent>;

  const createComponent = createComponentFactory({
    component: ResultComponent,
    imports: [provideTranslocoTestingModule({ en: {} })],
    detectChanges: false,
  });

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.debugElement.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('onRecommendationSelectedChange', () => {
    it('should set isRecommendedSelected', () => {
      component.isRecommendedSelected = true;

      component.onRecommendedSelectedChange(false);

      expect(component.isRecommendedSelected).toEqual(false);
    });
  });
});
