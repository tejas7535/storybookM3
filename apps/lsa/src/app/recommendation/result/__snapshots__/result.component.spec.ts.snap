// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultComponent when setting price and availability responses multiple times should set price and availability responses as an combined result 1`] = `
Object {
  "123456": Object {},
  "2222": Object {},
  "3333": Object {},
  "654321": Object {},
}
`;
